simulator/src/main.rs

```diff
@@ -0,0 +1,632 @@
+        DataEventType::GeneratePeripheralData => generate_peripheral_state_data(event.sequence),
+        DataEventType::GenerateMotor => generate_motor_data(event.sequence),
+        DataEventType::GenerateBMS => generate_bms_data(event.sequence),
+    };
+
+    let payload = serde_json::to_string(&data)?;
+    client.send(payload).await?;
+
+    let duration = event.event_type.duration();
+    let sequence = if event.sequence >= RESET_LIMIT { 1 } else { event.sequence + 1 };
+
+    events.insert(
+        Event::DataEvent(DataEvent {
+            sequence,
+            timestamp: event.timestamp + duration,
+            device: event.device.clone(),
+            event_type: event.event_type,
+        }),
+        duration,
+    );
+
+    Ok(())
+}
+
+async fn process_action_response_event(
+    event: &ActionResponseEvent,
+    client: &mut SplitSink<Framed<TcpStream, LinesCodec>, String>,
+) -> Result<(), Error> {
+    //info!("Sending action response {:?} {} {} {}", event.action_id, event.progress, event.status);
+    let payload = Payload {
+        stream: "action_status".to_string(),
+        device_id: None,
+        sequence: 0,
+        timestamp: SystemTime::UNIX_EPOCH.elapsed().unwrap().as_millis() as u64,
+        payload: json!({
+            "action_id": event.action_id,
+            "state": event.status,
+            "progress": event.progress,
+            "errors": []
+        }),
+    };
+
+    info!("Sending action response {} {} {}", event.action_id, event.progress, event.status);
+
+    let payload = serde_json::to_string(&payload)?;
+    client.send(payload).await?;
+    info!("Successfully sent action response");
+
+    Ok(())
+}
+
+pub async fn process_events(
+    events: &mut DelayQueue<Event>,
+    client: &mut SplitSink<Framed<TcpStream, LinesCodec>, String>,
+) -> Result<(), Error> {
+    if let Some(e) = events.next().await {
+        match e.into_inner() {
+            Event::DataEvent(event) => {
+                process_data_event(&event, events, client).await?;
+            }
+
+            Event::ActionResponseEvent(event) => {
+                process_action_response_event(&event, client).await?;
+            }
+        }
+    }
+
+    Ok(())
+}
+
+pub fn generate_action_events(action: Action, events: &mut DelayQueue<Event>) {
+    let action_id = action.action_id;
+
+    info!("Generating action events for action: {action_id}");
+    let now = Instant::now();
+
+    // Action response, 10% completion per second
+    for i in 1..10 {
+        let duration = Duration::from_secs(i as u64);
+        events.insert(
+            Event::ActionResponseEvent(ActionResponseEvent {
+                action_id: action_id.clone(),
+                progress: i * 10,
+                status: String::from("in_progress"),
+                timestamp: now + duration,
+            }),
+            duration,
+        );
+    }
+
+    let duration = Duration::from_secs(10);
+    events.insert(
```

Comment:
Shouldn't we use `insert_at` instead `insert` and check if events are being generated on time?